rfcs/0039-minaform-tool-design.md

```diff
@@ -0,0 +1,143 @@
+
+1) Production Networks vs. QA Nets
+    - The process for deploying production testnets vs. qa nets has diverged in a few places and this means our day-to-day testing does not cover all of the services we need to deploy for public networks.
+2) Determining network health is difficult/unclear/inconsistent
+    - Some of this is addressed with more health checks in k8s and more metrics / better dashboards in grafana, but integrating the deployment tooling with all of those checks can help make it more foolproof.
+3) No "uninstall" command / cleanup afterwards
+    - Both the testnet terraform files and resources in the cluster end up hanging around more than neccessary. Needs additional tooling to clean up after itself.
+4) Generating Genesis Proofs is slow/resource intensive/hard to predict
+    - See proposal for how to pre-generate proofs in docker images
+5) Hard to determine the correct docker tag from a given git SHA
+    - Can be scripted to allow for a git tag or git SHA input, in addition to / in place of the docker tag.
+6) Keys and other private info for deployment are not committed/shared with other team members
+    - Addressed by https://github.com/MinaProtocol/coda-automation/issues/472
+7) Incompatible helm/terraform versions
+    - Can be addressed by 1) checking versioning before trying to use these tools and 2) bundling a canonical set of tools into a docker image
+8) Incompatible helm/terraform changes
+    - Addressed by moving the charts to the mina repo and explicitly releasing them. Terraform specifies an individual version of each chart to depend on and will fetch the released chart before deploys.
+9) Helm string processing is inconsistent/hard
+    - Now all charts are passing `helm lint` in the mina repo, and we can add `helm test` in the future as well. Charts in the mina repo also allow for us to build dhall types and tooling around our charts, but that is out-of-scope for this RFC.
+10) Passing flags to the daemon process directly is difficult to debug/handle/manage
+    -  Expose command line flags genericly as a list of command line arguments in helm, allowing deployments to replace the default flags as needed and clearly specify them. These helm changes are just best-practices, and are also out-of-scope for this RFC.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+One of the major design goals is to introduce a set of default terraform network configurations for qa and production that will be the starting point for all networks we deploy, and will be maintained with PR's and strict code review, while still allowing any further modification to fit the needs of any individual deployment, testing, or debugging effort. The design revolves around a new bash script, `minaform`, as well improvements to the existing coda-network tool, that can be bundled in a container alongside its dependencies for easy cross-platform use as soon as possible.  Why bash? See #rationale-and-alternatives
```

Comment:
As mentioned by others, a bash script doesn't quite seem like the right tool here and that's not really taking into account the relatively nonexistent and overall poor writeability, readability and most importantly imo testability of bash (by infra + in a way all of engineering) when compared to other languages like `python, javascript` and others mentioned in comment.

Feels like the major goal here (introduce a set of default terraform network configurations for qa and production...) could be accomplished with a native `terraform` construct, *tfvars*, which allow expression of sets of variables exactly as expected within src and capable of being either statically defined or generated dynamically in a similar fashion to how the integration framework currently generates terraform code.